Bug: cleanup mass-soft-deletes library when CREATE TEMP TABLE _stash_scene_ids hits a different pooled Prisma connection

## Summary

`StashSyncService.cleanup`'s scene-deletion path uses `CREATE TEMP TABLE IF NOT EXISTS _stash_scene_ids` followed by `DELETE FROM`/`INSERT INTO`/`SELECT FROM` against that table. SQLite temp tables are **per-connection**, but Prisma uses a connection pool — so when the second statement lands on a different pooled connection, the temp table doesn't exist there. The `DELETE` raises `OperationalError: no such table: _stash_scene_ids`, the batched `INSERT`s also fail, and the final `SELECT … WHERE id NOT IN (SELECT id FROM _stash_scene_ids)` runs **against an empty table** — meaning every scene currently in the local DB is classified as "not in Stash" and soft-deleted.

In our environment this caused ~99.7% of scenes (122,883 of 123,209) to be silently flagged `deletedAt` in a single bulk event around 2026-02-03, with the only visible signal being a single ERROR line in the logs:

```
[ERROR] Failed to cleanup deleted scenes
{"error":"... Raw query failed. Code: `1`. Message: `no such table: _stash_scene_ids`"}
```

The error is logged but doesn't surface anywhere user-visible. The downstream `ExclusionComputationService` then classified ~95% of performers/studios/tags as `reason='empty'` (because they had no non-deleted scenes), making the entire library effectively invisible in the UI.

## Affected version

v3.3.6 (latest at time of report). Code path:

https://github.com/carrotwaxr/peek-stash-browser/blob/dc514a5/server/services/StashSyncService.ts#L55250-L57772 (around the `case "scene"` branch in the cleanup loop).

## Reproduction

The race depends on which connection the cleanup invocation lands on relative to the previous query, so it's intermittent. The conditions we can confirm:

- Library size: 100k+ scenes (forces many batched INSERTs, increasing chance of connection rotation)
- Concurrent FastAPI/HTTP read traffic on the DB during sync (forces pool to spread)
- Default Prisma pool sizing

When it fails, the symptoms are:
1. ERROR log line as above
2. `StashScene.deletedAt` populated for ~all rows in a single batch (matching `now` timestamp)
3. `UserExcludedEntity` floods with `reason='empty'` rows on next exclusion recompute
4. Homepage / browse view collapses to near-zero visible scenes

## Suggested fix

Pin the four operations to a single connection. Two viable paths:

**Option A — wrap in `prisma.$transaction([…])`**: forces a serial single-connection execution and is the smallest diff.

**Option B — use a regular (non-`TEMP`) table** keyed by an ephemeral run-id (pid + timestamp), with an `INSERT OR IGNORE`/`DELETE` lifecycle. Survives connection rotation and avoids the temp-table semantics entirely. The trade-off is a slightly bigger schema migration, but reads and writes from the pool become trivial.

Both options should also add a fail-loud guard around the `SELECT … NOT IN (...)` query — if the temp table is empty when it shouldn't be (e.g. `stashIds.length > 0 && SELECT count(*) FROM _stash_scene_ids == 0`), refuse to soft-delete and throw. This would have prevented our incident regardless of which mitigation lands.

## Recovery for affected users

For anyone who hits this and wants to recover their library:

```sql
-- 1. Snapshot first (sqlite3 .backup '...db.bak-<timestamp>')
-- 2. Stop the peek-stash-browser container so writes don't conflict
-- 3. Run, then restart container:
BEGIN IMMEDIATE;
UPDATE StashScene SET deletedAt = NULL WHERE deletedAt IS NOT NULL;
DELETE FROM UserExcludedEntity WHERE reason = 'empty';
COMMIT;
```

The first statement undoes the bulk soft-delete (Stash is canonical, so anything legitimately deleted in Stash will get re-flagged on next cleanup pass once the temp-table bug is fixed). The second clears the cascaded `empty` exclusions; the next exclusion recompute will repopulate any that are genuinely empty.

## Health canary

For users wanting an early-warning alarm:

```sql
SELECT COUNT(*) FROM StashScene WHERE deletedAt IS NOT NULL;
```

Healthy value is ~ the count of scenes you've actually deleted in Stash. If this jumps by thousands in a single cleanup cycle, the bug has fired.

---

Happy to send a PR for either Option A or Option B if it'd help.

--- a/server/integration/services/StashSyncService.cleanup.integration.test.ts
+++ b/server/integration/services/StashSyncService.cleanup.integration.test.ts
@@ -1,0 +1,133 @@
+/**
+ * Integration tests for StashSyncService.cleanupDeletedEntities (scene path).
+ *
+ * These run against the real test SQLite database (unlike the mocked unit tests
+ * in tests/services/StashSyncService.cleanup.test.ts) and exercise the actual
+ * temp-table `NOT IN` computation inside the single-connection transaction, plus
+ * the #526 delete-ratio safety threshold.
+ *
+ * Strategy: seed scenes under a dedicated, isolated stashInstanceId (so the real
+ * sync never touches them), stub getStashClient to return a controlled keep-set
+ * for that instance only, run cleanup, and assert exactly which scenes got
+ * soft-deleted in the real DB.
+ */
+import {
+  afterAll,
+  afterEach,
+  beforeEach,
+  describe,
+  expect,
+  it,
+  vi,
+} from "vitest";
+import prisma from "../../prisma/singleton.js";
+import { stashSyncService } from "../../services/StashSyncService.js";
+
+// Skip if no database connection (matches other integration tests).
+const describeWithDb = process.env.DATABASE_URL ? describe : describe.skip;
+
+// Isolated instance id — not a real synced Stash instance, so background sync
+// and other tests never touch these rows.
+const TEST_INSTANCE = "cleanup-it-instance";
+const SCENE_IDS = Array.from({ length: 10 }, (_, i) => String(i + 1)); // "1".."10"
+
+async function clearTestScenes(): Promise<void> {
+  await prisma.stashScene.deleteMany({
+    where: { stashInstanceId: TEST_INSTANCE },
+  });
+}
+
+async function seedTestScenes(): Promise<void> {
+  await clearTestScenes();
+  await prisma.stashScene.createMany({
+    data: SCENE_IDS.map((id) => ({ id, stashInstanceId: TEST_INSTANCE })),
+  });
+}
+
+/** Snapshot of which seeded scenes are alive vs soft-deleted. */
+async function snapshot(): Promise<{ alive: string[]; deleted: string[] }> {
+  const rows = await prisma.stashScene.findMany({
+    where: { stashInstanceId: TEST_INSTANCE },
+    select: { id: true, deletedAt: true },
+  });
+  return {
+    alive: rows.filter((r) => r.deletedAt === null).map((r) => r.id),
+    deleted: rows.filter((r) => r.deletedAt !== null).map((r) => r.id),
+  };
+}
+
+describeWithDb("StashSyncService.cleanupDeletedEntities (integration)", () => {
+  // The keep-set the stubbed Stash client reports for TEST_INSTANCE.
+  let keepSet: string[] = [];
+
+  beforeEach(async () => {
+    await seedTestScenes();
+
+    // Capture the real implementation (bound) before spying so calls for any
+    // OTHER instance pass through untouched — only TEST_INSTANCE gets the stub.
+    const realGetStashClient = (
+      stashSyncService as unknown as {
+        getStashClient: (id?: string) => unknown;
+      }
+    ).getStashClient.bind(stashSyncService);
+
+    vi.spyOn(
+      stashSyncService as unknown as {
+        getStashClient: (id?: string) => unknown;
+      },
+      "getStashClient"
+    ).mockImplementation((id?: string) => {
+      if (id !== TEST_INSTANCE) return realGetStashClient(id);
+      return {
+        findSceneIDs: async () => ({
+          findScenes: {
+            scenes: keepSet.map((sceneId) => ({ id: sceneId })),
+            count: keepSet.length,
+          },
+        }),
+      };
+    });
+  });
+
+  afterEach(() => {
+    vi.restoreAllMocks();
+  });
+
+  afterAll(async () => {
+    await clearTestScenes();
+  });
+
+  it("soft-deletes exactly the scenes missing from Stash (real temp-table NOT IN)", async () => {
+    // Stash still has scenes 1..7; 8, 9, 10 were deleted → 3/10 = 30% (under threshold).
+    keepSet = ["1", "2", "3", "4", "5", "6", "7"];
+
+    const deletedCount = await (
+      stashSyncService as unknown as {
+        cleanupDeletedEntities: (t: string, id: string) => Promise<number>;
+      }
+    ).cleanupDeletedEntities("scene", TEST_INSTANCE);
+
+    expect(deletedCount).toBe(3);
+    const { alive, deleted } = await snapshot();
+    expect(new Set(deleted)).toEqual(new Set(["8", "9", "10"]));
+    expect(new Set(alive)).toEqual(
+      new Set(["1", "2", "3", "4", "5", "6", "7"])
+    );
+  });
+
+  it("aborts and soft-deletes nothing when the keep-set looks truncated (#526 threshold)", async () => {
+    // Stash returns only 2 of 10 → would delete 8/10 = 80% (over the 50% threshold).
+    keepSet = ["1", "2"];
+
+    const deletedCount = await (
+      stashSyncService as unknown as {
+        cleanupDeletedEntities: (t: string, id: string) => Promise<number>;
+      }
+    ).cleanupDeletedEntities("scene", TEST_INSTANCE);
+
+    expect(deletedCount).toBe(0);
+    const { alive, deleted } = await snapshot();
+    expect(deleted).toHaveLength(0);
+    expect(new Set(alive)).toEqual(new Set(SCENE_IDS));
+  });
+});

--- a/server/services/StashSyncService.ts
+++ b/server/services/StashSyncService.ts
@@ -97,6 +97,13 @@ const ENTITY_PLURALS: Record<EntityType, string> = {
 // Constants for sync configuration
 const BATCH_SIZE = 500; // Number of entities to fetch per page
 
+// Cleanup safety: refuse to soft-delete more than this fraction of the live
+// (non-deleted) local entities in a single cleanup pass. A truncated or partial
+// ID list from Stash (e.g. a broken paginated fetch, or - per #526 - a temp table
+// landing on a different pooled connection) would otherwise mass-delete real
+// scenes. Soft-deletes are recoverable, but the guard stops it before it happens.
+const MAX_CLEANUP_DELETE_RATIO = 0.5;
+
 /**
  * Format a timestamp for Stash GraphQL queries.
  *
@@ -1375,6 +1382,39 @@ class StashSyncService extends EventEmitter {
   }
 
   /**
+   * Safety guard against a bad/truncated keep-set wiping a large share of the
+   * library. Returns true (and logs loudly) when the proposed soft-delete count
+   * exceeds MAX_CLEANUP_DELETE_RATIO of the live (non-deleted) local entities.
+   *
+   * This protects against ANY cause of a bad keep-set - a partial paginated
+   * fetch, a transient Stash error, or the connection-pool race described in
+   * #526 - not just one specific failure mode. The keep-set size is surfaced in
+   * the log so a truncated fetch is easy to spot.
+   */
+  private exceedsCleanupDeleteThreshold(
+    plural: string,
+    keepSetSize: number,
+    liveCount: number,
+    toDeleteCount: number
+  ): boolean {
+    // Nothing to protect (empty/new library) or nothing to delete - let it run.
+    if (liveCount === 0 || toDeleteCount === 0) {
+      return false;
+    }
+    const ratio = toDeleteCount / liveCount;
+    if (ratio > MAX_CLEANUP_DELETE_RATIO) {
+      logger.error(
+        `Cleanup safety: refusing to soft-delete ${toDeleteCount}/${liveCount} ${plural} ` +
+          `(${(ratio * 100).toFixed(1)}% > ${(MAX_CLEANUP_DELETE_RATIO * 100).toFixed(0)}% limit). ` +
+          `Stash returned only ${keepSetSize} ${plural} ID(s) - the list looks truncated or partial, ` +
+          `so cleanup is being skipped to prevent mass deletion. Run a full sync if this is expected.`
+      );
+      return true;
+    }
+    return false;
+  }
+
+  /**
    * Cleanup entities that no longer exist in Stash.
    * Called during full sync after each entity type has been synced.
    *
@@ -1514,31 +1554,54 @@ class StashSyncService extends EventEmitter {
           // 1. Loading all scene IDs into memory
           // 2. Exceeding SQLite's parameter limit (~32k)
           //
-          // Strategy: Insert stashIds into temp table, then query scenes NOT IN temp table
-
-          // Create temp table and populate with Stash IDs in batches
-          await prisma.$executeRawUnsafe(`CREATE TEMP TABLE IF NOT EXISTS _stash_scene_ids (id TEXT PRIMARY KEY)`);
-          await prisma.$executeRawUnsafe(`DELETE FROM _stash_scene_ids`);
-
-          const BATCH_SIZE = 500;
-          for (let i = 0; i < stashIds.length; i += BATCH_SIZE) {
-            const batch = stashIds.slice(i, i + BATCH_SIZE);
-            if (batch.length > 0) {
-              const values = batch.map((id) => `('${id}')`).join(",");
-              await prisma.$executeRawUnsafe(`INSERT OR IGNORE INTO _stash_scene_ids (id) VALUES ${values}`);
-            }
-          }
-
-          // Find scenes to delete (in local DB but not in Stash) - only fetch what we need
+          // The CREATE TEMP TABLE -> INSERT -> SELECT (NOT IN) -> DROP sequence MUST run on a
+          // single connection: a TEMP table is connection-scoped, so if the SELECT landed on a
+          // different pooled connection it would see an empty table and mark the entire library
+          // for deletion (#526). Peek pins connection_limit=1 today, but we run the whole
+          // sequence inside one interactive transaction so the guarantee holds explicitly
+          // regardless of pool/adapter configuration. The transaction is kept short - it only
+          // computes the delete-set; reconciliation and the soft-delete writes happen outside,
+          // after the safety threshold check below.
+          const sceneBatchSize = 500;
           const instanceFilter = stashInstanceId ? `stashInstanceId = '${stashInstanceId}'` : `stashInstanceId IS NULL`;
-          const scenesToDelete = await prisma.$queryRawUnsafe<Array<{ id: string; phash: string | null }>>(
-            `SELECT id, phash FROM StashScene
-             WHERE deletedAt IS NULL
-             AND ${instanceFilter}
-             AND id NOT IN (SELECT id FROM _stash_scene_ids)`
+          const scenesToDelete = await prisma.$transaction(
+            async (tx) => {
+              await tx.$executeRawUnsafe(`CREATE TEMP TABLE IF NOT EXISTS _stash_scene_ids (id TEXT PRIMARY KEY)`);
+              await tx.$executeRawUnsafe(`DELETE FROM _stash_scene_ids`);
+
+              for (let i = 0; i < stashIds.length; i += sceneBatchSize) {
+                const batch = stashIds.slice(i, i + sceneBatchSize);
+                if (batch.length > 0) {
+                  const values = batch.map((id) => `('${id}')`).join(",");
+                  await tx.$executeRawUnsafe(`INSERT OR IGNORE INTO _stash_scene_ids (id) VALUES ${values}`);
+                }
+              }
+
+              // Find scenes to delete (in local DB but not in Stash) - only fetch what we need
+              const rows = await tx.$queryRawUnsafe<Array<{ id: string; phash: string | null }>>(
+                `SELECT id, phash FROM StashScene
+                 WHERE deletedAt IS NULL
+                 AND ${instanceFilter}
+                 AND id NOT IN (SELECT id FROM _stash_scene_ids)`
+              );
+
+              await tx.$executeRawUnsafe(`DROP TABLE IF EXISTS _stash_scene_ids`);
+              return rows;
+            },
+            // Generous timeout: the insert loop can run many batches for very large libraries.
+            // Cleanup is a serial maintenance step, so a long-lived txn is fine.
+            { maxWait: 10000, timeout: 120000 }
           );
 
           if (scenesToDelete.length > 0) {
+            // Safety threshold (#526): abort before mutating anything if this would soft-delete
+            // an implausibly large share of the library, which indicates a truncated/partial
+            // keep-set rather than genuine deletions in Stash.
+            const liveCount = await this.getLocalEntityCount("scene", stashInstanceId);
+            if (this.exceedsCleanupDeleteThreshold(plural, stashIds.length, liveCount, scenesToDelete.length)) {
+              return 0;
+            }
+
             // Check for merges and reconcile user data before soft-deleting
             for (const scene of scenesToDelete) {
               if (scene.phash) {
@@ -1560,8 +1623,8 @@ class StashSyncService extends EventEmitter {
             // Soft-delete in batches
             const deleteIds = scenesToDelete.map((s) => s.id);
             const instanceId = stashInstanceId;
-            for (let i = 0; i < deleteIds.length; i += BATCH_SIZE) {
-              const batch = deleteIds.slice(i, i + BATCH_SIZE);
+            for (let i = 0; i < deleteIds.length; i += sceneBatchSize) {
+              const batch = deleteIds.slice(i, i + sceneBatchSize);
               await prisma.stashScene.updateMany({
                 where: { id: { in: batch }, stashInstanceId: instanceId },
                 data: { deletedAt: now },
@@ -1569,9 +1632,8 @@ class StashSyncService extends EventEmitter {
             }
             deletedCount = deleteIds.length;
           }
-
-          // Clean up temp table
-          await prisma.$executeRawUnsafe(`DROP TABLE IF EXISTS _stash_scene_ids`);
+          // Note: the _stash_scene_ids temp table is created and dropped inside the
+          // transaction above, so there is nothing to clean up here.
           break;
         }
         case "performer": {

--- a/server/tests/services/StashSyncService.cleanup.test.ts
+++ b/server/tests/services/StashSyncService.cleanup.test.ts
@@ -7,8 +7,8 @@
 import { describe, it, expect, beforeEach, afterEach, vi } from "vitest";
 
 // Mock prisma before any imports - define mock inline (vi.mock is hoisted)
-vi.mock("../../prisma/singleton.js", () => ({
-  default: {
+vi.mock("../../prisma/singleton.js", () => {
+  const mockPrisma: Record<string, unknown> = {
     stashScene: {
       updateMany: vi.fn().mockResolvedValue({ count: 0 }),
       findMany: vi.fn().mockResolvedValue([]),
@@ -73,8 +73,14 @@ vi.mock("../../prisma/singleton.js", () => ({
     $queryRaw: vi.fn(),
     $executeRawUnsafe: vi.fn().mockResolvedValue(undefined),
     $queryRawUnsafe: vi.fn().mockResolvedValue([]),
-  },
-}));
+  };
+  // Interactive transaction: invoke the callback with the same mock acting as `tx`
+  // so the scene cleanup's temp-table sequence runs against the mocked raw methods.
+  mockPrisma.$transaction = vi.fn(
+    async (cb: (tx: typeof mockPrisma) => unknown) => cb(mockPrisma)
+  );
+  return { default: mockPrisma };
+});
 
 // Create mock functions for stash API
 const mockFindSceneIDs = vi.fn();
@@ -161,10 +167,17 @@ vi.mock("../../utils/logger.js", () => ({
 // Import after mocking
 import { stashSyncService } from "../../services/StashSyncService.js";
 import prisma from "../../services/../prisma/singleton.js";
+import { mergeReconciliationService } from "../../services/MergeReconciliationService.js";
 
 describe("StashSyncService Cleanup", () => {
   beforeEach(() => {
     vi.clearAllMocks();
+    // Re-establish implementations that restoreAllMocks (afterEach) would wipe.
+    vi.mocked(prisma.$transaction).mockImplementation(
+      async (cb: unknown) => (cb as (tx: typeof prisma) => unknown)(prisma)
+    );
+    vi.mocked(prisma.$executeRawUnsafe).mockResolvedValue(undefined);
+    vi.mocked(prisma.$queryRawUnsafe).mockResolvedValue([]);
   });
 
   afterEach(() => {
@@ -397,6 +410,99 @@ describe("StashSyncService Cleanup", () => {
 
       expect(result).toBe(0);
       expect(prisma.stashScene.updateMany).not.toHaveBeenCalled();
+    });
+  });
+
+  describe("Cleanup safety threshold (#526)", () => {
+    it("should skip scene cleanup when the delete ratio exceeds the threshold", async () => {
+      // Stash returns a small, likely-truncated keep-set...
+      mockFindSceneIDs.mockResolvedValue({
+        findScenes: { scenes: [{ id: "1" }, { id: "2" }, { id: "3" }], count: 3 },
+      });
+      // ...the NOT IN query reports 80 local scenes would be soft-deleted...
+      vi.mocked(prisma.$queryRawUnsafe).mockResolvedValue(
+        Array.from({ length: 80 }, (_, i) => ({ id: `${i + 100}`, phash: null }))
+      );
+      // ...out of 100 live scenes locally -> 80% > 50% threshold.
+      vi.mocked(prisma.stashScene.count).mockResolvedValue(100);
+
+      const result = await (stashSyncService as any).cleanupDeletedEntities(
+        "scene",
+        "test-instance"
+      );
+
+      expect(result).toBe(0);
+      // Mass deletion blocked: no soft-deletes, and no merge reconciliation either.
+      expect(prisma.stashScene.updateMany).not.toHaveBeenCalled();
+      expect(mergeReconciliationService.reconcileScene).not.toHaveBeenCalled();
+    });
+
+    it("should proceed with scene cleanup when the delete ratio is under the threshold", async () => {
+      mockFindSceneIDs.mockResolvedValue({
+        findScenes: { scenes: [{ id: "1" }], count: 1 },
+      });
+      // Only 2 of 100 live scenes would be deleted -> 2% < 50%.
+      vi.mocked(prisma.$queryRawUnsafe).mockResolvedValue([
+        { id: "4", phash: null },
+        { id: "5", phash: null },
+      ]);
+      vi.mocked(prisma.stashScene.count).mockResolvedValue(100);
+      vi.mocked(prisma.stashScene.updateMany).mockResolvedValue({ count: 2 });
+
+      const result = await (stashSyncService as any).cleanupDeletedEntities(
+        "scene",
+        "test-instance"
+      );
+
+      expect(result).toBe(2);
+      expect(prisma.stashScene.updateMany).toHaveBeenCalled();
+    });
+
+    it("should not apply the threshold when there are no live scenes locally", async () => {
+      // Empty/new library: liveCount === 0, so the ratio guard must not divide-by-zero
+      // or block.
+      mockFindSceneIDs.mockResolvedValue({
+        findScenes: { scenes: [{ id: "1" }], count: 1 },
+      });
+      vi.mocked(prisma.$queryRawUnsafe).mockResolvedValue([
+        { id: "9", phash: null },
+      ]);
+      vi.mocked(prisma.stashScene.count).mockResolvedValue(0);
+      vi.mocked(prisma.stashScene.updateMany).mockResolvedValue({ count: 1 });
+
+      const result = await (stashSyncService as any).cleanupDeletedEntities(
+        "scene",
+        "test-instance"
+      );
+
+      expect(result).toBe(1);
+      expect(prisma.stashScene.updateMany).toHaveBeenCalled();
+    });
+  });
+
+  describe("Single-connection transaction (#526)", () => {
+    it("should run the temp-table cleanup sequence inside one transaction and drop the table", async () => {
+      mockFindSceneIDs.mockResolvedValue({
+        findScenes: { scenes: [{ id: "1" }, { id: "2" }], count: 2 },
+      });
+      // No scenes to delete - we only care that the temp-table sequence is wrapped.
+      vi.mocked(prisma.$queryRawUnsafe).mockResolvedValue([]);
+
+      await (stashSyncService as any).cleanupDeletedEntities(
+        "scene",
+        "test-instance"
+      );
+
+      // The whole CREATE/INSERT/SELECT/DROP sequence runs in a single interactive
+      // transaction so it is guaranteed to be same-connection (#526).
+      expect(prisma.$transaction).toHaveBeenCalledTimes(1);
+      const execSql = vi
+        .mocked(prisma.$executeRawUnsafe)
+        .mock.calls.map((c) => String(c[0]));
+      expect(execSql.some((sql) => /CREATE TEMP TABLE/i.test(sql))).toBe(true);
+      expect(
+        execSql.some((sql) => /DROP TABLE IF EXISTS _stash_scene_ids/i.test(sql))
+      ).toBe(true);
     });
   });
 });

